restli-tools/src/main/java/com/linkedin/restli/tools/snapshot/check/PegasusSchemaSnapshotCompatibilityChecker.java

```diff
@@ -0,0 +1,282 @@
+
+    if (summary.length() > 0)
+    {
+      System.out.println(summary);
+    }
+
+    if (cl.hasOption("report"))
+    {
+      System.out.println(new CompatibilityReport(compatibilityChecker._infoMap, compatLevel).createReport());
+      System.exit(0);
+    }
+
+    System.exit(compatibilityChecker._infoMap.isCompatible(compatLevel) ? 0 : 1);
+  }
+
+  /**
+   * Check backwards compatibility between a pegasusSchemaSnapshot (.pdl) and a pegasusSchemaSnapshot (.pdl) file.
+   *
+   * @param prevPegasusSchemaPath previously existing snapshot file
+   * @param currentPegasusSchemaPath current snapshot file
+   * @param compatibilityLevel compatibility level
+   * @param compatMode compatibilityOptions mode which defines the compatibility check mode.
+   * @return CompatibilityInfoMap which contains information whether the given two files are compatible or not.
+   */
+  public CompatibilityInfoMap checkPegasusSchemaCompatibility(String prevPegasusSchemaPath, String currentPegasusSchemaPath,
+      CompatibilityLevel compatibilityLevel, CompatibilityOptions.Mode compatMode)
+  {
+    _infoMap.isCompatible(compatibilityLevel);
+
+    File preSchemaFile = new File(prevPegasusSchemaPath);
+    File currSchemaFile = new File(currentPegasusSchemaPath);
+
+    DataSchema preSchema = null;
+    try
+    {
+      preSchema = parseSchema(preSchemaFile);
+    }
+    catch(FileNotFoundException e)
+    {
+      constructCompatibilityMessage(CompatibilityMessage.Impact.OLD_READER_IGNORES_DATA,
+          "New schema %s is created.", currentPegasusSchemaPath);
+    }
+
+    DataSchema currSchema = null;
+    try
+    {
+      currSchema = parseSchema(currSchemaFile);
+    }
+    catch(FileNotFoundException e)
+    {
+      constructCompatibilityMessage(CompatibilityMessage.Impact.BREAKS_OLD_READER,
+          "Schema %s is removed.", prevPegasusSchemaPath);
+    }
+
+    if (preSchema == null || currSchema == null)
+    {
+      return _infoMap;
+    }
+
+    CompatibilityOptions compatibilityOptions = new CompatibilityOptions().setMode(compatMode).setAllowPromotions(true);
+    CompatibilityResult result = CompatibilityChecker.checkCompatibility(preSchema, currSchema, compatibilityOptions);
+
+    if (!result.getMessages().isEmpty())
+    {
+      result.getMessages().forEach(message -> _infoMap.addModelInfo(message));
+    }
+
+    return _infoMap;
+  }
+
+  private void constructCompatibilityMessage(CompatibilityMessage.Impact impact, String format, Object... args)
+  {
+    CompatibilityMessage message = new CompatibilityMessage(new Object[]{}, impact, format, args);
+    _infoMap.addModelInfo(message);
+  }
+
+  private DataSchema parseSchema(File schemaFile) throws FileNotFoundException
+  {
+    DataSchemaResolver resolver = MultiFormatDataSchemaResolver.withBuiltinFormats(_resolverPath);
+    PdlSchemaParser parser = new PdlSchemaParser(resolver);
+    parser.parse(new FileInputStream(schemaFile));
```

Comment:
It seems odd that you need to create a resolver for each file. Do you want to process the same schema multiple times if it is referenced by other schemas? Consider this scenario:

Prev:
A.pdl:
record A {
  b: B
}
B.pdl:
record B {
  f1: int
  f2: float
}

Current:
A.pdl:
record A {
  b: B
}
B.pdl:
record B {
  f1: int
  // f2 deleted
}

Now will the error be reported once for B.pdl or twice, for both A.pdl and B.pdl?
I know we report the error multiple times for resource compatibility checking, but do you think that is the best options for schema checker also?

Reply:
Umm, I think based on the way we generated snapshot pdl we have to parse each file.
All the referenced schemas are constructed as inline records in the snapshot.
For example: 
snapshot: A.pdl will be look like following
record A {
  b: record B {
    f1: int 
    f2: float
  }
}

I don't think we could avoid it.